### ISSUE TITLE
Process the Clinician Order from Epic FHIR

### ISSUE THREAD

**Issue by aeyates** [OHSUCMP/coach#54]
On a patient's first login to COACH, we should be able to retrieve a ServiceRequest from FHIR representing the order from the physician to use COACH.

In the OHSU Epic instance:
Coding system:
POC: 1.2.840.114350.1.13.135.3.7.2.696580
PRD: 1.2.840.114350.1.13.135.2.7.2.696580

Code:
340941

If we need to capture a date to begin the monitoring block, use the "authoredOn" field provided in the resource.

In the note field, there should be a parseable blood pressure goal, something like:

```
"note": [
                    {
                        "text": "BP Goal\r\nSystolic: 130\r\nDiastolic: 80"
                    }
                ]
```

The goal should be parsed from the note and used when present to set the participant's goal in the trial. We also need to understand how this will work at other sites.

**Comment by aeyates** [comment]
We need to address a number of other questions around what the order means in terms of this "monitoring block", because we are starting to see requirements that COACH behave differently when the block is over:

- Will all participants have an active order? What if they don't (e.g., current patients at OHSU not in the trial). How are people not in the trial being handled in COACH? Are they getting redirected to REDCap to enroll like everyone else?
- When the order comes in, what happens with blood pressures collected before the order? Are we showing these on the graph? Are we using them to score and give recommendations? Implementation wise, it's a big lift to change how this works now to treat people with orders in a special way. As it stands now, a person entering COACH for the first time could already have met the criteria for having enough BP readings to finish monitoring.

**Comment by davedorr9** [comment]
An order will not be present at Missouri. So we need to build the system so it benefits from the order but isn't dependent on it.

If an order exists, then the system shall : parse the BP goal, **updating** it if the most recent updated goal is older than the order (or there is no updated BP goal).

If an order exists : submit all the BPs to Epic taken within the last month before the order comes in. Ok if these are duplicates, this will almost never happens

If an order comes in, all BPs before the BP are still valid.

If an order does not exist, then the system shall : use the default BP goal (140/90)

**Comment by mattStorer** [comment]
I worked on this for a few hours today, got some help from Dena in the process.  Trying to use Postman to get these Orders from POC to see what they look like, so what I code into COACH would work correctly.  In these tests, I was not able to get the correct types of orders.  Dena verified that she could get them on her end, but that was using a _provider_ context, **not** the _patient_ context, which is what COACH uses.

long story short, it sounds like COACH will not be able to get these orders with BP goals, at least when patients launch COACH via MyChart.

below is a screencap of some of my conversation with Dena that summarizes the issue pretty cleanly, copied and pasted with permission:

![image](https://github.com/OHSUCMP/coach/assets/979938/e81b9f2d-f317-4dc6-a93b-16f8e8dc439f)

**Comment by davedorr9** [comment]
If we launch the patient on COACH from the provider view, will that work?

**Comment by mattStorer** [comment]
> If we launch the patient on COACH from the provider view, will that work?

my understanding is that launching from the provider context, i.e. from Hyperspace, does work.

Amy also recently observed that using the backend context (using a JWT) also appears to bring back the desired Order types.

**Comment by mattStorer** [comment]
**DEVELOPER NOTE** : neither Logica nor Meld appear to permit _ServiceRequest_ resources to be created within their systems, as attempting to import generates an _OperationOutcome_ response with message "Failed to call access method: java.lang.NullPointerException".

this assessment is supported by the fact that the Patient Manager tool in each system displays a list of resources for the patient on the left, _whether or not the patient has any of those resources populated_.  that is, most of those resources show "(0)" beside them, indicating that the patient has 0 of those resources populated.  the implication here is that _every resource possibly available to be populated_ is listed there, and _ServiceRequest_ isn't one of them.

as such, it appears that it will be impossible to test this logic in any place except in POC.

black box testing it is.

**Comment by aeyates** [comment]
@mattStorer If you want to test outside of POC, the SmartHealth sandbox could be used. As a test, I added a patient with a service request (https://r4.smarthealthit.org/ServiceRequest?subject=Melinda). The sandbox gets wiped daily, so you have to add back the data each day.

**Comment by mattStorer** [comment]
@aeyates oh that's a thought.  good idea!  I'll chat you on Teams to discuss some details about that strategy, don't want to clog up the comment thread here.

**Comment by mattStorer** [comment]
tested changes to read a _ServiceRequest_ order and to convert it into a BP goal, if appropriate.  works as expected in local testing against the smarthealthit.org sandbox.  this local testing necessarily can only work using the _patient_ context, but we will need it to work using the _backend_ context in POC and PRD.

as such, at this point I feel confident that COACH can functionally perform the desired operation, so long as the _ServiceRequest_ resource can be acquired.  the next step will require black box testing in the POC environment where we can test its functionality using the _backend_ context.

**Comment by mattStorer** [comment]
this is the test patient I've been working with locally.  it contains a _ServiceRequest_ resource that looks like the type of resource we can expect to get from POC.  It references a BP goal of 133/88, which, when launching COACH for this patient, sets their initial BP goal to 133/88 and marks _bpGoalUpdated_=`true`

[OHSU-MBS-Test1.json](https://github.com/OHSUCMP/coach/files/13824683/OHSU-MBS-Test1.json)

**Comment by mattStorer** [comment]
just got this deployed to POC and tested there with the _Lilly Day Fhir_ user.  it appears that this is all working out of the box!

Lilly has the following BP goal of **135/90** defined in a _ServiceRequest_ resource (note: elements truncated for brevity, only showing "interesting" or otherwise relevant data):

```
{
	"resourceType": "ServiceRequest",
	"id": "e4Jnoi1aei.BW0bLlQQCUGaaD9r1Et1vU3KHcya6gZyc3",
	"status": "active",
	"intent": "filler-order",
	"category": [
	  {
		"coding": [
		  {
			"system": "urn:oid:1.2.840.114350.1.13.135.3.7.10.798268.30",
			"code": "4",
			"display": "Other"
		  }
		],
		"text": "Other"
	  }
	],
	"code": {
	  "coding": [
		{
		  "system": "urn:oid:1.2.840.114350.1.13.135.3.7.2.696580",
		  "code": "340941",
		  "display": "COACH PATIENT REPORTED BLOOD PRESSURE"
		}
	  ],
	  "text": "COACH PATIENT REPORTED BLOOD PRESSURE"
	},
	"subject": {
	  "reference": "Patient/eU5DTkzk8HnwId.Z5dIiN1mwHb2fBnvfIy9mQqNyrexc3",
	  "display": "Fhir, Lilly Day"
	},
	"authoredOn": "2023-12-21T23:01:45Z",
	"note": [
	  {
		"text": "BP Goal\r\nSystolic: 135\r\nDiastolic: 90"
	  }
	]
}
```

We can see in the logs that _ServiceRequest_ resources are pulled using the `BACKEND` context:

_fhirqueries.properties_:
```
ServiceRequest.Query=ServiceRequest?subject={subject}&category=4
ServiceRequest.Strategy=BACKEND
```

_coach.log_:
```
2024-01-04 06:56:47.612 [pool-2-thread-1] INFO  e.ohsu.cmp.coach.service.FHIRService [FHIRService.java:165] search: got Bundle with total=75, entries=76 for query: ServiceRequest?subject=eU5DTkzk8HnwId.Z5dIiN1mwHb2fBnvfIy9mQqNyrexc3&category=4
2024-01-04 06:56:47.618 [pool-2-thread-1] INFO  e.o.c.coach.workspace.UserWorkspace [UserWorkspace.java:420] DONE building Order ServiceRequests for session=D3019CDF196BD61B67DDD451C1CEA853 (size=18, took 1999ms)
2024-01-04 06:56:47.624 [http-nio-50000-exec-8] INFO  e.o.c.coach.workspace.UserWorkspace [UserWorkspace.java:439] DONE building remote Goals for session=D3019CDF196BD61B67DDD451C1CEA853 (size=2, took 2004ms)
```

and finally, we can see that the BP goal of 135/90 is displayed on the COACH home page after clearing supplemental data (which removes all goals in the COACH database, permitting the goal defined in the _ServiceRequest_ to be considered to be the most current and thusly displayed):

![image](https://github.com/OHSUCMP/coach/assets/979938/29c07e57-fbab-4b3d-a4bf-7a04dc76cc65)

so I think we're good as far as primary functionality goes with respect to this issue.

**Comment by mattStorer** [comment]
thoughts with respect to a comment above that appears to outline some additional requirements:

> If an order exists, then the system shall : parse the BP goal, updating it if the most recent updated goal is older than the order (or there is no updated BP goal).

if a _ServiceRequest_-based Order with a BP goal exists, the system parses the BP goal from its _note_ text and uses that as the current BP goal (I think this is what's meant by "updating it") until such time that a newer BP goal is identified (which could come from another _ServiceRequest_ or _Goal_ resource, **or** if the user configures their BP goal manually through COACH).

> If an order exists : submit all the BPs to Epic taken within the last month before the order comes in. Ok if these are duplicates, this will almost never happens

I'm not sure what this means, precisely.  Any BPs entered into COACH will be written back to Epic, irrespective of the presence of a _ServiceRequest_ Order BP goal, so I _think_ this is referring to adjusting the starting date that we will want to use for initial Omron synchronization to be one month before the _authoredOn_ date specified in the _ServiceRequest_ Order.  Is this understanding correct @davedorr9 ?

> If an order comes in, all BPs before the BP are still valid.

This doesn't make sense to me.  Please clarify.

> If an order does not exist, then the system shall : use the default BP goal (140/90)
COACH is configured to use a default goal of 140/90 in the absence of a _ServiceRequest_ or _Goal_ based BP goal, which will be created into the COACH database and used until such a time that a more current BP goal is identified.

**Comment by mattStorer** [comment]
I think that the requirement:

> If an order comes in, all BPs before the BP are still valid.

was meant to read _If an order comes in, all BPs before the **Order** are still valid_, in response to question by @aeyates:

> When the order comes in, what happens with blood pressures collected before the order? Are we showing these on the graph?

in which case, yes I confirm that COACH does nothing with BP readings in response to the presence of a _ServiceRequest_ Order.  They all remain valid, appear in the graph, and are incorporated into recommendations as usual.  All an Order does is potentially modify the user's BP goal (if it's the most current).

**Comment by mattStorer** [comment]
what if a _ServiceRequest_ Order with a BP goal is created _after_ a user has set a BP goal within COACH?  The way I have it configured right now, the current BP goal is whichever has the most recent effective date, irrespective of the goal source (whether it comes from a _Goal_ or _ServiceRequest_ resource originating from Epic, **or** from the local COACH database).

this makes sense to me - **current = most recent** - but let me know if we want to give a hard priority to, for example, BP goal created within COACH by the user.

**Comment by davedorr9** [comment]
# #

> if an order exists : submit all the BPs to Epic taken within the last month before the order comes in. Ok if these are duplicates, this will almost never happens

I'm not sure what this means, precisely. Any BPs entered into COACH will be written back to Epic, irrespective of the presence of a ServiceRequest Order BP goal, so I think this is referring to adjusting the starting date that we will want to use for initial Omron synchronization to be one month before the authoredOn date specified in the ServiceRequest Order. Is this understanding correct @davedorr9 ? />

##
If an order doesn't exist, any previous home blood pressures will fail at being written to Epic. Once an order is written, can the system go back 30 days and try to write all the home BPs? This is the same as the Omron requirement to go back one month, yes, but I'm not sure when that gets triggered - just trigger it when you get a new order

NEWEST goal = the correct goal, no need to guess which goal is right. If it is wrong, they can just update the goal again through coach. So if you get a new order with a goal, that should overwrite last BP goal in coach.
